<a href="https://colab.research.google.com/github/Living-with-machines/dhoxss-text2tech/blob/kb-hf/Sessions/5b-Transformers-for-NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NLP (and other stuff) with Hugging Face 🤗




Hugging Face is a company that, in its own words, is "on a mission to democratize good machine learning, one commit at a time." [link](https://huggingface.co/huggingface)

In broad terms, their goal is to make models and data more accessible to a wide range of users (also in the Humanities and the GLAM sector). 

Hugging Face is:
- a **Hub** for sharing deep learning models and data
- the developer of **software libraries**. We will focus on the `transformers` and `datasets` libraries.

**Q**: **What is a computational model?** Usually, by model we refer to a 'machine' that can perform certain (predictive) tasks (or any task in the context of AGI). It maps inputs to outputs, for example, text to sentiment predictions in the context of sentiment detection.

The `transformers` library has become one of the most popular libraries for NLP:
* State-of-the-art models are easier to use.
* Provides APIs to download and use pre-trained models, but also allows you to load and fine-tune your models.
* It is open source!

`transfomers` allows you to interact with the **Model Hub**: a central point for people to share and find models. They host more than 815K models, supporting different languages and tasks.


The `datasets` library offers tools for sharing and manipulating large amounts of data for machine learning. It also allows you to download data others have prepared for you, which is often very handy. The Hub currently contains more than 189K datasets.

We will only scratch the surface, but if you are interested in this, we highly recommend the HuggingFace course: https://huggingface.co/course

Let's look at the [models page](https://huggingface.co/models)...

You can:
- Search by name
- Navigate by Task (Text Classification and Token Classification)

You can consult specific **model cards** to know more about:
- the popularity of a model
- the datasets  and base-models used for fine-tuning

For example:
- [distilbert-base-uncased-finetuned-sst-2-english](https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english)
- [bert-base-NER](https://huggingface.co/dslim/bert-base-NER)

Please note:
- model pages often contain **widgets** that allow you to experiment and get a sense of their behaviour (but you'd need to login to play with them).

... and the [datasets](https://huggingface.co/datasets).

If you are interested in data and models for humanities research, we recommend having a closer look at [BigLAM](https://huggingface.co/biglam).

## Hugging Face and spaCy?

- Both implement Transformer models for NLP
- spaCy is more convenient for processing text using robust NLP models
- Hugging Face focuses more on machine learning applications and has more flexibility

✏️ **Exercises:**
    
* Explore the model hub: https://huggingface.co/models.
* Have a look at the models: tasks and languages available.
* Discussion:
    * Can you find a model that fits your needs (language, task, ...)?
    * Have you spotted any strange behaviour?

## The `transformers` pipelines

As we have seen, the Hugging Face Hub provides you with MANY models and datasets. But how to use them?

The main library is called `transformers`.
 and it provides a convenient way of using deep learning models for some of the main tasks in natural language processing (but also for computer vision and other domains if that's your thing).

The most convenient tool is the `pipeline` function from the `transformers` library. This tool packages many steps and makes working with a wide range of often complex models much easier.

A pipeline is a code object that abstracts away most of the complex code (it happens in the background), leaving only the bare minimum for the user to interact.

First, we need to install the `transformers` library. Run the code below to do this.

### Installing the transformers library

In [1]:
!pip install -q torch torchvision torchaudio sentencepiece
!pip install -q transformers datasets


[notice] A new release of pip is available: 23.2.1 -> 24.3.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.2.1 -> 24.3.1
[notice] To update, run: pip install --upgrade pip


Pipelines provide a simplified and uniform tool to use deep learning models for different tasks.


We load the `pipeline` module from the `transformers` library:

In [2]:
import transformers, torch
from transformers import pipeline # import the library

# Determine the device based on platform and availability
if torch.backends.mps.is_available() and torch.device("mps").type == "mps":  # macOS MPS
    device = torch.device("mps")
elif torch.cuda.is_available():  # Linux/Windows CUDA
    device = torch.device("cuda")
else:  # Fallback to CPU
    device = torch.device("cpu")

print(f"Using device: {device}")

/Users/fnanni/Projects/dhoxss-text2tech/.venv/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Using device: mps


Now, we can use the pipeline for several tasks:
- Sentiment Analysis
- Zero-shot classification
- Named Entity Recognition

... and more!

### Sentiment analysis

The task of determining whether a text is positive, negative, or neutral.


In [3]:
classifier = pipeline("sentiment-analysis",device=device) # instantiate a sentiment classifier
text = "Life is wonderful." # save a sentence in the text variable
classifier(text) # apply classifier to the sentence

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/Users/fnanni/Projects/dhoxss-text2tech/.venv/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


[{'label': 'POSITIVE', 'score': 0.9998849630355835}]

### Exercise

Change the sentence and compute the sentiment. Make it difficult (or British), like "the weather is not too bad."

In [4]:
# enter code here

###  A closer look at pipelines



In [5]:
from transformers import pipeline

To create a pipeline, you need to know:
* The **task** you want to perform (e.g. `'sentiment-analysis'`). To know the currently accepted tasks, you can inspect the values for the `task` parameter [here](https://huggingface.co/docs/transformers/main_classes/pipelines)
* The **model** you want to use to make predictions (e.g. `'distilbert-base-uncased'`), which must be trained for the task you want to perform (i.e. `fill-mask`).

In some scenarios it might be important to pay attention to:
* The tokenizer used by the model (i.e. the strategy that BERT uses to split sequences into smaller units. This is often the same name as the model, e.g. `'distilbert-base-uncased'`).
* The conventions the (language) model follows: e.g. if your task is `fill-mask` (which predicts the masked token), how is the masked element tagged (usually `[MASK]`, sometimes `<MASK>`, etc.).

If you obtained your model from the HuggingFace model hub (https://huggingface.co/models), you should be able to find all this info in the model card (e.g. https://huggingface.co/bert-base-uncased).

**Note:** You can find more information on pipelines and how to use them in https://huggingface.co/transformers/main_classes/pipelines.html.

In [6]:
# let's implement a multilingual sentiment analysis model
classifier = pipeline("sentiment-analysis",
                      model="nlptown/bert-base-multilingual-uncased-sentiment",
                      tokenizer="nlptown/bert-base-multilingual-uncased-sentiment",device=device)
classifier('Ik ben zo ongelukkig!')

[{'label': '1 star', 'score': 0.6801469922065735}]

Notice how the output changes, depending on the type of model you use (ultimately this depends on the training data).

### Exercise

Go to the Hugging Face Hub, find a model that performs sentiment detection for Italian and test it!


In [7]:
# enter code here

#### Zero-shot classification

The task of classifying texts into categories of your choice.

In [8]:
classifier = pipeline("zero-shot-classification",device=device) # instantiate a zero-shot classifier
classifier(
    "This is a course about the text mining in the humanities.",
    candidate_labels=["technology", "history", "business", "sports"],
)

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


{'sequence': 'This is a course about the text mining in the humanities.',
 'labels': ['history', 'technology', 'business', 'sports'],
 'scores': [0.4423690140247345,
  0.24387599527835846,
  0.18639598786830902,
  0.1273590475320816]}

#### Named entity recognition

The task of recognizing named entities in a text.

In [9]:
ner = pipeline("ner", grouped_entities=True,device=device)
ner("Federico, Kaspar and Mariona are in Oxford")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
/Users/fnanni/Projects/dhoxss-te

[{'entity_group': 'PER',
  'score': 0.99825984,
  'word': 'Federico',
  'start': 0,
  'end': 8},
 {'entity_group': 'PER',
  'score': 0.98075575,
  'word': 'Kaspar',
  'start': 10,
  'end': 16},
 {'entity_group': 'PER',
  'score': 0.9907509,
  'word': 'Mariona',
  'start': 21,
  'end': 28},
 {'entity_group': 'LOC',
  'score': 0.9962345,
  'word': 'Oxford',
  'start': 36,
  'end': 42}]

The previous pipelines provide classification tools at either the text or token level. Hugging Face also provides a few generative pipelines, i.e. pipelines that are not focused on labelling but on producing textual outputs.

We have a closer look at Generative AI in a second, but to show you a few examples.

#### Text generation

The task of generating text given a prompt.

In [10]:
generator = pipeline("text-generation",device=device) # instantiate a text generator
generator("It was the best of times")

No model was supplied, defaulted to openai-community/gpt2 and revision 6c0e608 (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'It was the best of times. I had fun in my life, my family."\n\nStill, he remembers telling his mom on the phone that she wanted more of him, and so the whole family came together.\n\n"It was the'}]

#### Machine translation

The task of converting a source text from one language to another.

In [11]:
translator = pipeline("translation", model="Helsinki-NLP/opus-mt-en-es",device=device)
translator("The inklings informally gathered at the Eagle and Child pub on Tuesday mornings.")

/Users/fnanni/Projects/dhoxss-text2tech/.venv/lib/python3.9/site-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


[{'translation_text': 'Los inklings se reunieron informalmente en el pub Eagle and Child los martes por la mañana.'}]

✏️ **Exercise:**

In [12]:
# Have a look at the other pipelines, get inspration from the tutorial:
# https://huggingface.co/course/chapter1/3?fw=pt#working-with-pipelines
# Play a bit with different pipelines, using different models from the
# model hub: https://huggingface.co/models (make sure that the language
# is trained for the task you would like to try!).
#
# Type your code here:



✏️ **Exercise:**

In [13]:
# Can you combine all the tools you have learned so far?
# Find a Spanish dataset on Hugging Face
# Translate all the texts to English
# Apply English NER to the translation
#
# Type your code here:



## Applying a pipeline to a data set

In [14]:
import pandas as pd
from datasets import Dataset
from transformers import pipeline
from tqdm.auto import tqdm

In [15]:
# download a sample of 10.000 newspaper articles
!wget -q --show-progress https://github.com/kasparvonbeelen/lancaster-newspaper-workshop/raw/wc/data/sample_lwm_hmd_mt90_10000.csv.zip
# unzip the downloaded sample
!unzip -o sample_lwm_hmd_mt90_10000.csv.zip
!rm -r __MACOSX

sample_lwm_hmd_mt90 100%[===================>]  37.90M  16.0MB/s    in 2.4s    
Archive:  sample_lwm_hmd_mt90_10000.csv.zip
  inflating: sample_lwm_hmd_mt90_10000.csv  
  inflating: __MACOSX/._sample_lwm_hmd_mt90_10000.csv  


In [16]:
df_small = pd.read_csv('sample_lwm_hmd_mt90_10000.csv').sample(3)
df_small.head(3)

,NLP,issue,art_num,title,collection,full_date,year,month,day,location,word_count,ocrquality,text,decade
5955,3408,807,art0075,"The Dewsbury Chronicle, and West Riding Advert...",British Library Living with Machines Project,1895-08-07,1895,8,7,"Dewsbury, West Yorkshire, England",1794,0.9678,"school, and he offered the new mistress the he...",1890
6609,3061,705,art0024,Brighouse & Rastrick Gazette.,British Library Living with Machines Project,1890-07-05,1890,7,5,"Brighouse, West Yorkshire, England",1512,0.9794,LADIES' COLUMN. The London season is nearing ...,1890
8056,3040,922,art0094,The Birkenhead News and Wirral General Adverti...,British Library Living with Machines Project,1894-09-22,1894,9,22,"Birkenhead, Merseyside, England",845,0.9576,PUBLIC-HOUSE MANAGERS AND THEIR VOTES. __ Rob...,1890


In [17]:


labels = ['politics','economy','crime','foreign affairs'] # define the labels
dataset = Dataset.from_pandas(df_small) # convert the dataframe to a dataset
hypothesis_template = "This text is about {}" # define the hypothesis template
classifier = pipeline("zero-shot-classification",
                      model="MoritzLaurer/deberta-v3-large-zeroshot-v2.0",device=device) # instantiate a zero-shot classifier

def classify(row):
    out = classifier(
              row["text"], # select the text column
              candidate_labels=labels,
              hypothesis_template=hypothesis_template,
              )
    return {'labels': out['labels'], 'scores': out['scores']} 

dataset = dataset.map(classify) # apply the classify function to the dataset

Map: 100%|██████████| 3/3 [00:06<00:00,  2.05s/ examples]


In [18]:
dataset[0]['text'][:100]

"school, and he offered the new mistress the hearty congratulation of the Governors upon her year's w"

In [19]:
dataset[0]['labels'], dataset[0]['scores']

(['foreign affairs', 'politics', 'economy', 'crime'],
 [0.3158459961414337,
  0.2588949203491211,
  0.22724813222885132,
  0.19801098108291626])

# Advanced Materials (Optional)

## Case Study: Using Historical Language Models

### 1. Static word embeddings

Introduced in 2013, word2vec has had a huge impact in natural language processing and its applications.

Vector representations of words seem to capture word meaning quite well!

Accessible and easy to use (easy to train, to apply and to share).

Shortcoming: this algorithm creates static embeddings, i.e. it creates one vector per word, no matter how many meanings the word has (e.g. `I like apples` vs `I like Apple macbooks`.)

Import the `gensim` library:

In [20]:
import gensim
import gensim.downloader

Download and load one of the models.

Just for illustration, we'll use `glove-wiki-gigaword-50`, which was trained on text from Wikipedia and Gigaword (newswire). Note that different models may perform differently.

In [23]:
glove_vectors = gensim.downloader.load('glove-wiki-gigaword-50')

[==================================================] 100.0% 66.0/66.0MB downloaded


Static word embeddings create one vector per word.

_Example 1:_
See top 20 most similar words to word 'mouse'. What do you observe?

In [24]:
glove_vectors.most_similar('mouse', topn=20)

[('monkey', 0.7965500950813293),
 ('bugs', 0.7805659174919128),
 ('cat', 0.773166835308075),
 ('rabbit', 0.7622703313827515),
 ('worm', 0.750491201877594),
 ('clone', 0.7307788729667664),
 ('robot', 0.7268993854522705),
 ('spider', 0.7199547290802002),
 ('bug', 0.7104865312576294),
 ('frog', 0.7027060389518738),
 ('mice', 0.6978598833084106),
 ('morph', 0.6821691989898682),
 ('rat', 0.6796256303787231),
 ('ape', 0.677627444267273),
 ('monster', 0.6768894791603088),
 ('click', 0.6624072194099426),
 ('uses', 0.6548233032226562),
 ('squirrel', 0.6545741558074951),
 ('creature', 0.6523107886314392),
 ('trackball', 0.6500849723815918)]

_Example 2:_ See top 20 most similar words to word 'pear'.

What would you expect to see here? Any guesses on the top most similar? And what do you see?

In [25]:
glove_vectors.most_similar('pear', topn=20)

[('mango', 0.854493260383606),
 ('avocado', 0.8034905195236206),
 ('pineapple', 0.7993020415306091),
 ('pomegranate', 0.7979760766029358),
 ('apricot', 0.7934627532958984),
 ('plum', 0.7909936308860779),
 ('peach', 0.7899217009544373),
 ('tomato', 0.7809722423553467),
 ('almond', 0.7721766829490662),
 ('guava', 0.7694117426872253),
 ('cherries', 0.7693701386451721),
 ('cucumber', 0.7676603198051453),
 ('cherry', 0.7642222046852112),
 ('melon', 0.7587339282035828),
 ('rhubarb', 0.7586070895195007),
 ('cranberry', 0.7529084086418152),
 ('papaya', 0.7493483424186707),
 ('compote', 0.7489796876907349),
 ('fruit', 0.7483691573143005),
 ('ripe', 0.7483317852020264)]

_Example 3:_ See top 20 most similar words to word 'apple'.

What would you expect to see here? Any guesses on the top most similar? And what do you see?

In [26]:
glove_vectors.most_similar('apple', topn=20)

[('blackberry', 0.7543067932128906),
 ('chips', 0.7438643574714661),
 ('iphone', 0.7429664134979248),
 ('microsoft', 0.7334205508232117),
 ('ipad', 0.7331036925315857),
 ('pc', 0.7217226624488831),
 ('ipod', 0.7199784517288208),
 ('intel', 0.7192243337631226),
 ('ibm', 0.7146540880203247),
 ('software', 0.7093585729598999),
 ('macintosh', 0.7047759890556335),
 ('android', 0.7046630382537842),
 ('processor', 0.6996651291847229),
 ('product', 0.6925289034843445),
 ('dell', 0.6896464228630066),
 ('cola', 0.6863355040550232),
 ('desktop', 0.6860974431037903),
 ('netscape', 0.6852995753288269),
 ('processors', 0.6781534552574158),
 ('amd', 0.6766293048858643)]

Word2Vec (2013) was one of the most important developments in NLP:
* It captures word meaning quite well!
* Very accessible and easy to use

... but it produces "static embeddings", with one vector representation for each word form.

**But most words have multiple meanings!**

### 2. Contextualized word embeddings

Words mean different things in different contexts.

**Goal:** learn the representation (i.e. the "meaning"!) for each word in its context.

In recent years (since 2018 mostly), lots of progress has been made (from BERT to GPT-3).

Also, lots of progress in making this easily accessible, and easy to use. The company HuggingFace has been greatly responsible for this last point, especially with their `transformers` library and their model hub.

A **transformer** is a deep learning model that uses the **attention** mechanism (a mechanism which is based on cognitive attention, and which focuses on where the key information in a sequence is produces while forgetting less relevant information). Its development has had a huge impact in deep learning, especially in natural language processing and computer vision. It allows a more effective modeling of long term dependencies between the words in a sequence, and more efficient training, not limited by the sequence order of the input sequence.

**BERT** (Bidirectional Encoder Representations from Transformers) is a transformer-based model, hugely successful, that creates contextualized word embeddings, it captures fine-grained contextual properties of words. It learns contextualized information through a masking process (i.e. it hides some words and uses their position to infer them back).

![](https://github.com/Living-with-machines/dhoxss-text2tech/blob/kb-hf/Sessions/images/mask_predictions.png?raw=1)

### 2.2. The Mask filling pipeline

Masked language modeling is the task of masking tokens in a sequence with a masking token, and prompting the model to fill that mask with an appropriate token (source: https://huggingface.co/transformers/task_summary.html#masked-language-modeling). The `fill-mask` pipeline replaces the mask in a sequence by the most likely prediction according to a BERT model.

We will create a `fill-mask` pipeline using the `distilbert-base-uncased` English model (and its tokenizer), as follows:

In [27]:
unmasker = pipeline('fill-mask',
                    model='distilbert-base-uncased',
                    tokenizer='distilbert-base-uncased',device=device)

/Users/fnanni/Projects/dhoxss-text2tech/.venv/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


This pipeline allows us to easily use BERT to predict the masked element in a sentence.

In the previous cell, we are:
* Creating a pipeline for the task of `fill-mask`,
* by using the `distilbert-base-uncased` BERT model and tokenizer,
* and storing the resulting pipeline in a variable (we call it `unmasker`), which we can use and reuse in subsequent code.

**Warning:** You need to make sure the model you use is trained for the `'fill-mask'` task.

To use the pipeline, you just need to pass the sentence containing the masked word as an argument of `unmasker` (i.e. the variable containing your pipeline). You don't need to do any encoding, the pipeline already takes care of converting the text into an input BERT can understand!

We store the output of applying the pipeline to this sentence in the `outputs` variable, as shown below:

In [28]:
outputs = unmasker("The cell is guarded by a [MASK].")

Now, let's inspect the `outputs` variable:

In [29]:
print(outputs)

[{'score': 0.1419995129108429, 'token': 3457, 'token_str': 'guard', 'sequence': 'the cell is guarded by a guard.'}, {'score': 0.09074389189481735, 'token': 8638, 'token_str': 'fence', 'sequence': 'the cell is guarded by a fence.'}, {'score': 0.05313451588153839, 'token': 2813, 'token_str': 'wall', 'sequence': 'the cell is guarded by a wall.'}, {'score': 0.04400521516799927, 'token': 10684, 'token_str': 'keeper', 'sequence': 'the cell is guarded by a keeper.'}, {'score': 0.037186313420534134, 'token': 4796, 'token_str': 'gate', 'sequence': 'the cell is guarded by a gate.'}]


In [30]:
# Let's print the results in an easier-to-read format:
for one_output in outputs:
    print("Prediction:", one_output['token_str'])
    print("Score:     ", round(one_output['score'],4))
    print()

Prediction: guard
Score:      0.142

Prediction: fence
Score:      0.0907

Prediction: wall
Score:      0.0531

Prediction: keeper
Score:      0.044

Prediction: gate
Score:      0.0372



In [31]:
outputs = unmasker("""When a cell has been produced, we can then trace some of the
                      stages by which new [MASK] are formed. There appear to be four
                      modes in which vegetable cells are multiplied. The new cells
                      may either proceed from a nucleus or they may be formed at
                      once in the protoplasm.""")

# Let's print the results in an easier-to-read format:
for one_output in outputs:
    print("Prediction:", one_output['token_str'])
    print("Score:     ", round(one_output['score'],4))
    print()

Prediction: cells
Score:      0.9774

Prediction: nuclei
Score:      0.0077

Prediction: neurons
Score:      0.0026

Prediction: colonies
Score:      0.0017

Prediction: organisms
Score:      0.0012



In [32]:
outputs = unmasker("""Imprisonment with proper employment, and at least two visits
                      every day from a prison officer. The punishment does not
                      extend over a month. A week must elapse before the same
                      prisoner can be put again into the dark [MASK].""")

# Let's print the results in an easier-to-read format:
for one_output in outputs:
    print("Prediction:", one_output['token_str'])
    print("Score:     ", round(one_output['score'],4))
    print()

Prediction: ##room
Score:      0.1958

Prediction: cell
Score:      0.0785

Prediction: tower
Score:      0.0615

Prediction: room
Score:      0.0517

Prediction: chamber
Score:      0.0445



### Working historical models

To study meaning change over time, we can apply `fill-mask` to historical language models.

In [33]:
unmasker_1760_1850 = pipeline('fill-mask',
                              model='Livingwithmachines/bert_1760_1850',
                              tokenizer='Livingwithmachines/bert_1760_1850',device=device)

Some weights of the model checkpoint at Livingwithmachines/bert_1760_1850 were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [34]:
# We can now use them to predict a mask in a sentence as well:
outputs = unmasker_1760_1850("""The [MASK] is guarded by guards.""")

# Let's print the results in an easier-to-read format:
for one_output in outputs:
    print("Prediction:", one_output['token_str'])
    print("Score:     ", round(one_output['score'],4))
    print()

Prediction: door
Score:      0.1674

Prediction: entrance
Score:      0.0997

Prediction: gate
Score:      0.0459

Prediction: house
Score:      0.0369

Prediction: hall
Score:      0.036



✏️ **Exercise:**

In [ ]:
# Create a pipeline for the 1890-1900 model as well and try different sentences with
# both the 1760-1850 and the 1890-1900 models. Do language models trained on data
# from different periods make different predictions?
#
# Type your code here:



### Case Study: Vector representations over time


The meaning of a word, in NLP, is usually represented as vectors, i.e. lists of numbers. In the case of transformer models (such as BERT), the vector of a word changes based on the context in which this word occurs. In the following cells, we'll see how to obtain the vector representation of a token using the `tranformers` library.

#### Tokenization

As we saw yesterday, tokenizing a text is splitting it into meaningful units. Very often, this means splitting a text into words.

BERT uses a **subword** tokenization procedure called WordPiece.

This means that it does not only separate words. It also splits certain words into (ideally) meaningful units
> E.g. it splits the word `tokenizing` into `token` and `##izing`, where `##` indicates that this is a suffix which should be attached to the previous word).

These tokens are then mapped to numbers.

The tokenizer maps every word form (e.g. `token` and `##izing`) with identifiers in the vocabulary (e.g. given a certain model, `19204` is the vocabulary ID of `token` and `6026` is the vocabulary ID of the suffix `##izing`, so the word 'tokenizing' would be tokenized as `[19204, 6026]`).

**!!! Warning:** BERT has certain limits as to the length of the string that is accepted, which depends on the model, but usually 512 tokens.

**!!! VERY important:** different models may have different token-to-id mappings. When we use an existing model, we must use the same tokenizer (and therefore the same vocabulary mapping) that was used when training the model.

#### The inner workings of BERT tokenization

Tokenization steps:

1. The text is split into tokens, which can be:
  * words
  * parts of words
  * punctuation symbols

2. The tokenizer adds special tokens:
  * `[CLS]` indicating that this is the beginning of the input sequence.
  * `[SEP]` indicating that it is the end of the sequence (or a sequence delimiter if we have a pair of sequences as input).

3. The tokenizer maps each token into their vocabulary IDs.

Let's explore this in code.

In [35]:
# Load the tokenizer of a certain BERT model
our_tokenizer = transformers.AutoTokenizer.from_pretrained("distilbert-base-uncased")

In [36]:
# The `encode` function does the three steps in one go (splits the sequence, adds
# special tokens, and converts them into a sequence of IDs):
encoded_seq = our_tokenizer.encode('The cell is relentlessly guarded by guards.')
print(encoded_seq)

[101, 1996, 3526, 2003, 21660, 2135, 13802, 2011, 4932, 1012, 102]


And there are also functions that translate the vocabulary IDs to the word forms (given a certain tokenizer):

In [37]:
# The `convert_ids_to_tokens` returns the tokens that correspond to the IDs of
# an encoded sequence:
tokens = our_tokenizer.convert_ids_to_tokens(encoded_seq)
print(tokens)

['[CLS]', 'the', 'cell', 'is', 'relentless', '##ly', 'guarded', 'by', 'guards', '.', '[SEP]']


### 2.6. The feature extraction pipeline

Here we will see how to get vectors for words in context.

Similarly to what we did with word2vec, we may also want to have access to the vector of a certain word. However, unlike with word2vec, the vector of a word will depend on the context in which the word occurs. This means that we can't just ask for the vector of the word "apple", for example: we will need to ask for the vector of the word "apple" given a certain context.

We first import the following two libraries, which will help us work with vectors:

In [38]:
import numpy as np # python library used for working with vectors
from scipy import spatial # package to help compute distance or similarity between vectors

The pipeline task to obtain the vectors for tokens in a sequence is `feature-extraction`. As you can see, creating this pipeline is very similar to creating the `fill-mask` pipeline.

We will store the pipeline in a variable called `nlp_features`:

In [39]:
nlp_features = pipeline("feature-extraction",
                    model='distilbert-base-uncased',
                    tokenizer='distilbert-base-uncased',device=device)

Given a sentence, the pipeline tokenizes the input sentence:

In [40]:
sentence = "They were told that the machines stopped working."

output = nlp_features(sentence)
output_vectors = np.squeeze(output) # This removes single-dimensional entries (i.e. for vector readability)

Let's inspect the output. First of all, let's print it:

In [41]:
print(output_vectors)

[[-0.0118581   0.08130863  0.01319036 ... -0.00107469  0.4374842
   0.30174702]
 [ 0.24342713  0.07197712 -0.00515071 ...  0.0965031   0.77282101
  -0.27633172]
 [ 0.36821371 -0.05984684  0.09242672 ... -0.00808026  0.3218767
  -0.06213425]
 ...
 [ 0.51063424  0.18607342 -0.05728513 ... -0.45785472  0.22284982
  -0.2392973 ]
 [ 0.60650522  0.13031702 -0.40193444 ...  0.03722487 -0.30919343
  -0.45531228]
 [ 0.17005569  0.39926389  0.36847538 ... -0.03275453  0.28973925
  -0.15749149]]


This is an array (a list of vectors). Let's see its shape:

In [42]:
print(output_vectors.shape) # Print the shape of the vector

(11, 768)


This means that we have an arrray (in other words a matrix, a table) that has 11 vectors of length 768 (or, in other words, 11 rows with 768 columns).

**Question:** 11 vectors? Why 11?

Let's see how the sentence is tokenized (we've seen how above):

In [43]:
# Load the **SAME** tokenizer used in the pipeline:
our_tokenizer = transformers.AutoTokenizer.from_pretrained("distilbert-base-uncased")

# Encode the sentence into a sequence of vocabulary IDs
encoded_seq = our_tokenizer.encode(sentence)
print(encoded_seq)

# And get the tokens given the vocabulary IDs
tokens = our_tokenizer.convert_ids_to_tokens(encoded_seq)
print(tokens)

# And print the length of the tokenized sequence:
print(len(tokens))

[101, 2027, 2020, 2409, 2008, 1996, 6681, 3030, 2551, 1012, 102]
['[CLS]', 'they', 'were', 'told', 'that', 'the', 'machines', 'stopped', 'working', '.', '[SEP]']
11


As you can see, the input sentence has been tokenized into 11 tokens. So what we have in the above array is 11 vectors (each one representing a word in the context of the sentence, **keeping the order of tokens**, i.e. the first vector will correspond to the special token `[CLS]`, the second vector to the token `the`, and so on until the last vector, which corresponds to the special token `[SEP]`).

How do we get the vector of a specific token?

In [44]:
print(tokens[6]) # The 6th element in the tokenized sentence is the token `machine` (we start counting from zero)

machines


In [45]:
print(output_vectors[6]) # Therefore, o the 6th vector in output_vectors is the vector of `machine` in this context.

[ 5.68442464e-01  2.85227299e-01 -2.64460742e-02  4.56377983e-01
  8.09054077e-01  2.10991234e-01 -1.86342120e-01  4.00198489e-01
 -5.87703288e-01 -1.16031297e-01  1.50383502e-01  1.17749222e-01
 -1.04229577e-01  4.52154905e-01 -3.61544043e-01 -9.57690552e-02
 -2.47727036e-01 -3.43580663e-01 -1.24239072e-01  6.80379689e-01
 -1.00477755e-01 -6.53509051e-02  2.04908833e-01  3.81003320e-01
 -4.68232065e-01 -4.58845720e-02 -2.61937737e-01 -4.10879776e-03
  1.30466968e-02  9.54242982e-03  5.04743636e-01  3.07510316e-01
 -1.05667682e-02  2.77194083e-02  3.43943805e-01 -9.40639079e-02
 -5.28898597e-01  2.94373482e-01 -2.21306711e-01 -4.47421148e-02
 -2.48401940e-01 -6.16603255e-01 -2.59287953e-02 -3.22031915e-01
  3.78465772e-01  3.45407903e-01 -4.24682736e-01 -3.16346705e-01
  2.02660695e-01 -3.62806290e-01 -3.84560049e-01  1.81433141e-01
  5.44739187e-01 -1.69176787e-01  1.86413586e-01  5.20396590e-01
 -4.48541403e-01 -5.73056899e-02  1.27630100e-01 -1.58515990e-01
 -2.01199576e-01  2.72798

✏️ **Exercise:**

In [ ]:
# Create two `feature-extraction` pipelines, one for the  1760-1850 model, and
# one for the 1890-1900 model. Find whether the cosine similarity between words
# in sequences change depending on which BERT model you use.
#
# Type your code here:

👀 **If you are interested in knowing more, we recommend:**
* The HuggingFace tutorials: https://huggingface.co/course/chapter1/1
* BERT for Humanists: http://www.bertforhumanists.org/